cpu/stm32/stmclk/stmclk_l0l1.c

```diff
@@ -22,67 +22,166 @@
+#define CLOCK_ENABLE_HSE            1
+#else
+#define CLOCK_ENABLE_HSE            0
+#endif
+
+/* Check whether HSI must be enabled:
+  - When HSI is used as SYSCLK
+  - When PLL is used as SYSCLK and the board doesn't provide HSE (since HSI will be
+    used as PLL input clock)
+*/
+#if IS_ACTIVE(CONFIG_USE_CLOCK_HSI) || \
+    (!IS_ACTIVE(CONFIG_BOARD_HAS_HSE) && IS_ACTIVE(CONFIG_USE_CLOCK_PLL))
+#define CLOCK_ENABLE_HSI            1
+#else
+#define CLOCK_ENABLE_HSI            0
+#endif
+
+/* Check whether MSI must be enabled:
+  - When MSI is used as SYSCLK
+*/
+#if IS_ACTIVE(CONFIG_USE_CLOCK_MSI)
+#define CLOCK_ENABLE_MSI            1
 #else
-#error "Please provide CLOCK_HSI or CLOCK_HSE in boards/NAME/includes/perhip_cpu.h"
+#define CLOCK_ENABLE_MSI            0
 #endif
 
 /**
  * @brief Configure the controllers clock system
  *
- * The clock initialization make the following assumptions:
- * - the external HSE clock from an external oscillator is used as base clock
- * - the internal PLL circuit is used for clock refinement
- *
- * Use the following formulas to calculate the needed values:
- *
- * SYSCLK = ((HSE_VALUE / CLOCK_PLL_M) * CLOCK_PLL_N) / CLOCK_PLL_P
- * USB, SDIO and RNG Clock =  ((HSE_VALUE / CLOCK_PLL_M) * CLOCK_PLL_N) / CLOCK_PLL_Q
- *
- * The actual used values are specified in the board's `periph_conf.h` file.
- *
- * NOTE: currently there is not timeout for initialization of PLL and other locks
+ * NOTE: currently there is no timeout for initialization of PLL and other locks
  *       -> when wrong values are chosen, the initialization could stall
  */
 void stmclk_init_sysclk(void)
 {
-    /* Reset the RCC clock configuration to the default reset state(for debug purpose) */
-    /* Set MSION bit */
-    RCC->CR |= RCC_CR_MSION;
-    /* Reset SW, HPRE, PPRE1, PPRE2, MCOSEL and MCOPRE bits */
-    RCC->CFGR &= ~(RCC_CFGR_PLLSRC | RCC_CFGR_PLLDIV | RCC_CFGR_PLLMUL);
-    /* Reset HSION, HSEON, CSSON and PLLON bits */
-    RCC->CR &= ~(RCC_CR_HSION | RCC_CR_HSEON | RCC_CR_HSEBYP | RCC_CR_CSSON | RCC_CR_PLLON);
+    /* disable any interrupts. Global interrupts could be enabled if this is
+     * called from some kind of bootloader...  */
+    unsigned is = irq_disable();
     /* Disable all interrupts */
+    REG_CIR = 0x0;
 
-#if defined(CPU_FAM_STM32L0)
-    RCC->CICR = 0x0;
-#elif defined(CPU_FAM_STM32L1)
-    RCC->CIR = 0x0;
-#else
-#error unexpected MCU
-#endif
+    /* enable HSI clock for the duration of initialization */
+    stmclk_enable_hsi();
+
+    /* Reset the RCC clock configuration to the default reset state(for debug purpose) */
+    /* Reset MSION, HSEON, CSSON and PLLON bits */
+    RCC->CR &= ~(RCC_CR_MSION | RCC_CR_HSEON | RCC_CR_HSEBYP | RCC_CR_CSSON | RCC_CR_PLLON);
+
+    /* use HSI as system clock while we do any further configuration and
+    * configure the AHB and APB clock dividers as configured by the board */
+    RCC->CFGR = (RCC_CFGR_SW_HSI | CLOCK_AHB_DIV | CLOCK_APB1_DIV | CLOCK_APB2_DIV);
```

Comment:
Why move when this is done?

Reply:
I don't understand this question. What do you mean ?

Reply:
Before the dividers where set later on in the initializtion process, I'm not saying it maters, just asking.

Reply:
Here an initial content is written at once and later only the bitfileds that need to be changed are modified. The AHB/APB prescalers can be set during this first batch, no need to reset them again later.